File: .github/scripts/make-post.js

import fs from 'node:fs';
import path from 'node:path';

const EVENT_PATH = process.env.GITHUB_EVENT_PATH;
if (!EVENT_PATH || !fs.existsSync(EVENT_PATH)) {
    console.error('GITHUB_EVENT_PATH not found.');
    process.exit(1);
}

const event = JSON.parse(fs.readFileSync(EVENT_PATH, 'utf8'));
const issue = event.issue || {};
const title = (issue.title || 'Untitled').trim();
const body  = (issue.body  || '').trim();

const pad = n => String(n).padStart(2, '0');
const now  = new Date(new Date().toLocaleString('en-GB', { timeZone: 'Europe/London' }));
const today = `${pad(now.getDate())}${pad(now.getMonth()+1)}${now.getFullYear()}`;

//date override: "date: DDMMYYYY" or "date: YYYY-MM-DD"
const m1 = body.match(/^\s*date:\s*(\d{2})(\d{2})(\d{4})\s*$/im);
const m2 = body.match(/^\s*date:\s*(\d{4})-(\d{2})-(\d{2})\s*$/im);
const ddmmyyyy = m1 ? `${m1[1]}${m1[2]}${m1[3]}`
    : m2 ? `${pad(m2[3])}${pad(m2[2])}${m2[1]}`
    : today;

// slug override: "slug: my-custom-slug"
const ms = body.match(/^\s*slug:\s*([a-z0-9-]+)\s*$/im);
const baseSlug = title.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/^-+|-+$/g, '') || 'post';
const slug = (ms ? ms[1] : baseSlug).toLowerCase();

const fileName = `${ddmmyyyy}-${slug}.md`;
const relPath  = path.join('blog', fileName);

// Simple content: H1 + body
const content = `# ${title}\n\n${body}\n`;
fs.mkdirSync('blog', { recursive: true });
fs.writeFileSync(relPath, content, 'utf8');

// Emit outputs to workflow
const out = process.env.GITHUB_OUTPUT;
if (out) {
    fs.appendFileSync(out, `path=${relPath}\n`);
    fs.appendFileSync(out, `title=${title}\n`);
    fs.appendFileSync(out, `date=${ddmmyyyy}\n`);
}

console.log(`Created ${relPath}`);

File: .github/scripts/make-post.py
#!/usr/bin/env python3
import os, re, json, sys, pathlib, datetime

EVENT_PATH = os.environ.get("GITHUB_EVENT_PATH")
if not EVENT_PATH or not os.path.exists(EVENT_PATH):
    print("GITHUB_EVENT_PATH not found.", file=sys.stderr)
    sys.exit(1)

with open(EVENT_PATH, "r", encoding="utf-8") as f:
    event = json.load(f)

issue = event.get("issue", {}) or {}
title = (issue.get("title") or "Untitled").strip()
body  = (issue.get("body")  or "").strip()

now = datetime.datetime.now()
dd = f"{now.day:02d}"
mm = f"{now.month:02d}"
yyyy = f"{now.year:04d}"
ddmmyyyy = f"{dd}{mm}{yyyy}"

slug = re.sub(r"[^a-zA-Z0-9 -]", "", title).strip().lower()
slug = re.sub(r"\s+", "-", slug) or "post"

rel_dir = "blog"
pathlib.Path(rel_dir).mkdir(parents=True, exist_ok=True)

base_name = f"{ddmmyyyy}-{slug}.md"
rel_path = os.path.join(rel_dir, base_name)
counter = 2
while os.path.exists(rel_path):
    rel_path = os.path.join(rel_dir, f"{ddmmyyyy}-{slug}-{counter}.md")
    counter += 1

content = f"# {title}\n\n{body}\n"
with open(rel_path, "w", encoding="utf-8") as f:
    f.write(content)

out = os.environ.get("GITHUB_OUTPUT")
if out:
    with open(out, "a", encoding="utf-8") as f:
        f.write(f"path={rel_path}\n")
        f.write(f"title={title}\n")
        f.write(f"date={ddmmyyyy}\n")

print(f"Created {rel_path}")